File: Algorithms/0092_Reverse_Linked_List_II/Python/Reverse_Linked_List_II_Solution_1.py

# Space: O(n)
# Time: O(n)

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:

    def reverseBetween(self, head, left, right):
        if head is None or head.next is None: return head
        if left == right: return head
        first, first_prev, second, second_next = None, None, None, None
        count = 1
        cur = head
        while cur is not None:
            if count == left - 1: first_prev = cur
            if count == left: first = cur
            if count == right: second = cur
            if count == right + 1:
                second_next = cur
                break
            count += 1
            cur = cur.next

        second.next = None
        reversed_list = self.reverse_list(first)

        if first_prev: first_prev.next = reversed_list
        else: head = reversed_list

        if second_next:
            temp_cur = head
            while temp_cur.next:
                temp_cur = temp_cur.next
            temp_cur.next = second_next

        return head

    def reverse_list(self, head):
        if head is None: return None
        if head.next is None: return head
        prev = self.reverse_list(head.next)
        head.next.next = head
        head.next = None
        return prev

    # def reverse_list1(self, head):
    #     if head is None: return None
    #     if head.next is None: return head
    #     prev, cur = None, head
    #     while cur:
    #         temp = cur.next
    #         prev = cur
    #         cur = temp
    #     return prev

File: Algorithms/0092_Reverse_Linked_List_II/Java/ReverseLinkedListIISolution1.java

// Space: O(n)
// Time: O(n)

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode reverseBetween(ListNode head, int left, int right) {
        if (head == null || head.next == null) return head;
        if (left == right) return head;
        ListNode firstPrev = null, first = null, second = null, secondNext = null, cur = head;
        int count = 1;
        while (cur != null && secondNext == null) {
            if (count == left - 1) firstPrev = cur;
            if (count == left) first = cur;
            if (count == right) second = cur;
            if (count == right + 1) secondNext = cur;
            cur = cur.next;
            count++;
        }
        second.next = null;

        if (firstPrev != null) firstPrev.next = reverseList1(first);
        else head = reverseList1(first);
        if (secondNext != null) {
            ListNode tempCur = head;
            while (tempCur.next != null) tempCur = tempCur.next;
            tempCur.next = secondNext;
        }
        return head;
    }

    private ListNode reverseList(ListNode root) {
        if (root == null || root.next == null) return root;
        ListNode temp = reverseList(root.next);
        root.next.next = root;
        root.next = null;
        return temp;
    }

    // private ListNode reverseList1(ListNode root) {
    //     if (root == null || root.next == null) return root;
    //     ListNode cur = root;
    //     ListNode prev = null;
    //     while (cur != null) {
    //         ListNode temp = cur.next;
    //         cur.next = prev;
    //         prev = cur;
    //         cur = temp;
    //     }
    //     return prev;
    // }
}




